src/main.rs

```diff
@@ -94,6 +138,14 @@ impl manager::NotificationWatcher for Program {
     }
 }
 
+fn print_controller(controller: &controller::Controller) {
```

Comment:
Do you think this could be the `Display` trait implementation for Controller instead ?